{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "\n",
    "def load_data(path):\n",
    "        \n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data, training_set_number):\n",
    "    \n",
    "    \n",
    "    if t_data == \"train\":\n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((30000,101,4))          \n",
    "    \n",
    "    elif t_data == \"test\":    \n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((10000,101,4))      \n",
    "    \n",
    "    \n",
    "    \n",
    "    r = 0    \n",
    "    \n",
    "   \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        \n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data,training_set_number):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "            \n",
    "        cobinding = np.zeros((30000,101,cobinding_data.shape[1]/101),dtype=np.int)    \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "        cobinding = np.zeros((10000,101,cobinding_data.shape[1]/101),dtype=np.int)\n",
    "   \n",
    "    \n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data, training_set_number):\n",
    "    \n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1)\n",
    "        region = np.zeros((30000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1) \n",
    "        region = np.zeros((10000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "    \n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n",
    "\n",
    "\n",
    "def get_fold (protein, t_data, training_set_number):\n",
    "    \n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((30000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((10000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "    \n",
    "    \n",
    "    return fold\n",
    "\n",
    "\n",
    "def load_data_sources(protein, t_data, training_set_number, *args):\n",
    "\n",
    "    X = np.array([])\n",
    "    data_sources = []\n",
    "    for arg in args:\n",
    "        \n",
    "        if arg == 'KMER':\n",
    "            if X.size == 0:\n",
    "                X = get_seq(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_seq(protein, t_data, training_set_number)))\n",
    "        if arg == 'RNA': \n",
    "            if X.size == 0:\n",
    "                X = get_fold(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_fold(protein, t_data, training_set_number)))\n",
    "        if arg == 'RG':   \n",
    "            if X.size == 0:\n",
    "                X = get_region(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_region(protein, t_data, training_set_number)))\n",
    "        if arg == 'CLIP': \n",
    "            if X.size == 0:\n",
    "                X = get_cobinding(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_cobinding(protein, t_data, training_set_number)))\n",
    "        data_sources.append(arg)\n",
    "        \n",
    "    data_sources = ' '.join(data_sources)     \n",
    "    return data_sources, X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KMER RNA RG CLIP\n",
      "(30000, 101, 36)\n"
     ]
    }
   ],
   "source": [
    "name, X = load_data_sources('1_PARCLIP_AGO1234_hg19', 'train', 0, 'KMER', 'RNA', 'RG', 'CLIP')\n",
    "\n",
    "print (name)\n",
    "print (X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello\n",
      "seq\n",
      "arg2 seq\n"
     ]
    }
   ],
   "source": [
    "def test_args_kwargs(*args):\n",
    "    name = []    \n",
    "    for arg in args:\n",
    "        if arg == 'arg2': print('hello')\n",
    "        if arg == 'seq' : print ('seq')\n",
    "        name.append(arg)\n",
    "    print(' '.join(name))\n",
    "#example_list_of_args\n",
    "test_args_kwargs('arg2','seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KMER RNA RG CLIP\n",
      "(30, 101, 4)\n"
     ]
    }
   ],
   "source": [
    "def get_fold (protein, t_data, training_set_number):\n",
    "    \n",
    "    return np.ones((30,101,1))\n",
    "\n",
    "def get_seq (protein, t_data, training_set_number):\n",
    "    \n",
    "    return np.ones((30,101,1))\n",
    "def get_region (protein, t_data, training_set_number):\n",
    "    \n",
    "    return np.ones((30,101,1))\n",
    "def get_cobinding (protein, t_data, training_set_number):\n",
    "    \n",
    "    return np.ones((30,101,1))\n",
    "\n",
    "\n",
    "def load_data_sources(protein, t_data, training_set_number, *args):\n",
    "\n",
    "    X = np.array([])\n",
    "    data_sources = []\n",
    "    for arg in args:\n",
    "        \n",
    "        if arg == 'KMER':\n",
    "            if X.size == 0:\n",
    "                X = get_seq(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_seq(protein, t_data, training_set_number)))\n",
    "        if arg == 'RNA': \n",
    "            if X.size == 0:\n",
    "                X = get_fold(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_fold(protein, t_data, training_set_number)))\n",
    "        if arg == 'RG':   \n",
    "            if X.size == 0:\n",
    "                X = get_region(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_region(protein, t_data, training_set_number)))\n",
    "        if arg == 'CLIP': \n",
    "            if X.size == 0:\n",
    "                X = get_cobinding(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_cobinding(protein, t_data, training_set_number)))\n",
    "        data_sources.append(arg)\n",
    "        \n",
    "    data_sources = ' '.join(data_sources)     \n",
    "    return data_sources, X\n",
    "\n",
    "\n",
    "data_sources, X = load_data_sources('1_PARCLIP_AGO1234_hg19', 'train', 0, 'KMER', 'RNA', 'RG', 'CLIP')\n",
    "print (data_sources)\n",
    "print (X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\n"
     ]
    }
   ],
   "source": [
    "#30k datasets \n",
    "\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.callbacks import ModelCheckpoint, EarlyStopping\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "\n",
    "def load_data(path):\n",
    "        \n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data, training_set_number):\n",
    "    \n",
    "    \n",
    "    if t_data == \"train\":\n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((30000,101,4))          \n",
    "    \n",
    "    elif t_data == \"test\":    \n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((10000,101,4))      \n",
    "    \n",
    "    \n",
    "    \n",
    "    r = 0    \n",
    "    \n",
    "   \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        \n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data,training_set_number):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "            \n",
    "        cobinding = np.zeros((30000,101,cobinding_data.shape[1]/101),dtype=np.int)    \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "        cobinding = np.zeros((10000,101,cobinding_data.shape[1]/101),dtype=np.int)\n",
    "   \n",
    "    \n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data, training_set_number):\n",
    "    \n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1)\n",
    "        region = np.zeros((30000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1) \n",
    "        region = np.zeros((10000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "    \n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n",
    "\n",
    "\n",
    "def get_fold (protein, t_data, training_set_number):\n",
    "    \n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((30000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((10000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "    \n",
    "    \n",
    "    return fold\n",
    "\n",
    "def load_data_sources(protein, t_data, training_set_number, *args):\n",
    "\n",
    "    X = np.array([])\n",
    "    data_sources = []\n",
    "    for arg in args:\n",
    "        \n",
    "        if arg == 'KMER':\n",
    "            if X.size == 0:\n",
    "                X = get_seq(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_seq(protein, t_data, training_set_number)))\n",
    "        if arg == 'RNA': \n",
    "            if X.size == 0:\n",
    "                X = get_fold(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_fold(protein, t_data, training_set_number)))\n",
    "        if arg == 'RG':   \n",
    "            if X.size == 0:\n",
    "                X = get_region(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_region(protein, t_data, training_set_number)))\n",
    "        if arg == 'CLIP': \n",
    "            if X.size == 0:\n",
    "                X = get_cobinding(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_cobinding(protein, t_data, training_set_number)))\n",
    "        data_sources.append(arg)\n",
    "        \n",
    "    data_sources = ' '.join(data_sources)     \n",
    "    return data_sources, X\n",
    "\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "table_data = {}\n",
    "dataset_list = ['KMER', 'RNA', 'RG', 'CLIP']\n",
    "training_set_number = 0\n",
    "\n",
    "# We have to generate combinations with differnet lengths, thus we generate length intigers and create sets with all combinations at given length.\n",
    "for length in range(1,5):\n",
    "    for set in itertools.combinations(dataset_list, length):\n",
    "\n",
    "        average_score = np.zeros(len(protein_list))\n",
    "        score_list = []\n",
    "\n",
    "        for protein in protein_list:\n",
    "           \n",
    "            data_sources, X = load_data_sources(protein, 'train', training_set_number, *set)\n",
    "            y = get_class(protein,\"train\",training_set_number)\n",
    "            size = X.shape[2]\n",
    "            score = []\n",
    "\n",
    "            model = Sequential()\n",
    "            model.add(Conv1D(10,6, data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "            model.add(MaxPooling1D(pool_size=20, strides=1, padding='valid'))\n",
    "            model.add(Conv1D(10, 4, activation='relu'))\n",
    "            model.add(Dropout(0.1))\n",
    "            model.add(MaxPooling1D(pool_size=40, strides=1, padding='valid'))\n",
    "            model.add(Conv1D(15, 4, activation='relu'))\n",
    "            model.add(MaxPooling1D(pool_size=30, strides=1, padding='valid'))\n",
    "            model.add(Conv1D(15, 3, activation='relu'))\n",
    "            model.add(GlobalAveragePooling1D())\n",
    "            model.add(Dense(200, activation='relu'))\n",
    "            model.add(Dropout(0.1))\n",
    "            model.add(Dense(2, activation='softmax'))\n",
    "\n",
    "            model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "            checkpointer = ModelCheckpoint(filepath=\"models/\" + protein + \"_weights.hdf5\", verbose=0, save_best_only=True)\n",
    "            earlystopper = EarlyStopping(monitor='val_loss', patience=5, verbose=0)\n",
    "\n",
    "            X_train = X\n",
    "            y_train = y\n",
    "\n",
    "            metrics = model.fit(X_train, y_train, validation_split = 0.2, epochs=12, batch_size=200, verbose=0, callbacks=[earlystopper])\n",
    "\n",
    "            #Run predictions on test dataset.\n",
    "            data_sources, X_test = load_data_sources(protein, 'test', 0, *set)\n",
    "            y_test = get_class(protein,\"test\",training_set_number)\n",
    "\n",
    "\n",
    "            y_scores = model.predict(X_test)\n",
    "            y_scores = y_scores [:,0:1]\n",
    "            y_test = y_test [:,0:1]\n",
    "\n",
    "            score = roc_auc_score(y_test, y_scores)\n",
    "\n",
    "            \n",
    "            score_list.append(score)\n",
    "        \n",
    "        print (protein)\n",
    "        table_data[data_sources] = score_list\n",
    "\n",
    "\n",
    "table = pd.DataFrame(table_data, index = [i+1 for i in range (len(protein_list))])\n",
    "table = table.T\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open('auc_combinations_results.pickle', 'wb') as f:\n",
    "    pickle.dump(table_data, f, protocol=pickle.HIGHEST_PROTOCOL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "def highlight_max(s):\n",
    "    is_max = s == max(s)\n",
    "    return ['background-color: yellow' if v else '' for v in is_max]\n",
    "\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "with open('auc_combinations_results.pickle', 'rb') as f:\n",
    "    loaded_table_data = pickle.load(f)\n",
    "\n",
    "for dict_value in loaded_table_data:\n",
    "    loaded_table_data[dict_value] = np.round(np.array(loaded_table_data[dict_value]),3)\n",
    "\n",
    "    \n",
    "loaded_table = pd.DataFrame(loaded_table_data, index = [i+1 for i in range (len(protein_list))])\n",
    "loaded_table = loaded_table.T\n",
    "loaded_table = loaded_table.reindex(sorted(loaded_table.columns), axis=1)\n",
    "\n",
    "loaded_table.style.apply(highlight_max)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import OrderedDict\n",
    "\n",
    "\n",
    "with open('auc_combinations_results.pickle', 'rb') as f:\n",
    "    loaded_table_data = pickle.load(f)\n",
    "    for source in loaded_table_data.keys():\n",
    "        std = stats.sem(loaded_table_data[source])\n",
    "        loaded_table_data[source] = np.true_divide(sum(loaded_table_data[source]),31)\n",
    "        print(u'AUC for {} is {} \\u00B1 {}'.format(source, np.round(loaded_table_data[source], 2), np.round(std, 2)))\n",
    "        \n",
    "\n",
    "    sorted_data = OrderedDict(sorted(loaded_table_data.items(), key=lambda x: x[1]))\n",
    "    \n",
    "    plt.bar(sorted_data.keys(), sorted_data.values(), align='edge')\n",
    "    plt.xticks(rotation='vertical')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from scipy.stats.stats import pearsonr  \n",
    "import pandas as pd\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "\n",
    "with open('auc_combinations_results.pickle', 'rb') as f:\n",
    "    loaded_table_data = pickle.load(f)\n",
    "\n",
    "    index = ['KMER' , 'RNA', 'RG', 'CLIP']\n",
    "    df = pd.DataFrame(index = index)\n",
    "\n",
    "    for source in index:\n",
    "        correlation = [(pearsonr(loaded_table_data[source],loaded_table_data[s]))[0] for s in index]\n",
    "        print(correlation)\n",
    "        df1 = pd.DataFrame(correlation, index = index, columns =  [source])\n",
    "        df = df.join(df1)\n",
    "\n",
    "    print(df)\n",
    "\n",
    "    h = sns.heatmap(df, cmap = sns.color_palette(\"Blues\"), annot=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "individual_sources = np.zeros((4,31))\n",
    "with open('auc_combinations_results.pickle', 'rb') as f:\n",
    "    loaded_table_data = pickle.load(f)\n",
    "    individual_sources = {k:loaded_table_data[k] for k in ['KMER' , 'RNA', 'RG', 'CLIP']}\n",
    "    g = sns.clustermap(individual_sources, method = 'ward', col_cluster = False, cmap = sns.cubehelix_palette(8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#30k datasets \n",
    "\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.callbacks import ModelCheckpoint, EarlyStopping\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "\n",
    "def load_data(path):   \n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data, training_set_number): \n",
    "    if t_data == \"train\":\n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((30000,101,4))          \n",
    "    \n",
    "    elif t_data == \"test\":    \n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((10000,101,4))      \n",
    "      \n",
    "    r = 0    \n",
    "      \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        \n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data,training_set_number):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "            \n",
    "        cobinding = np.zeros((30000,101,cobinding_data.shape[1]/101),dtype=np.int)    \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "        cobinding = np.zeros((10000,101,cobinding_data.shape[1]/101),dtype=np.int)\n",
    "   \n",
    "    \n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1)\n",
    "        region = np.zeros((30000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1) \n",
    "        region = np.zeros((10000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "    \n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n",
    "\n",
    "\n",
    "def get_fold (protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((30000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((10000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "    \n",
    "    \n",
    "    return fold\n",
    "\n",
    "def load_data_sources(protein, t_data, training_set_number, *args):\n",
    "    X = np.array([])\n",
    "    data_sources = []\n",
    "    for arg in args:\n",
    "        \n",
    "        if arg == 'KMER':\n",
    "            if X.size == 0:\n",
    "                X = get_seq(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_seq(protein, t_data, training_set_number)))\n",
    "        if arg == 'RNA': \n",
    "            if X.size == 0:\n",
    "                X = get_fold(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_fold(protein, t_data, training_set_number)))\n",
    "        if arg == 'RG':   \n",
    "            if X.size == 0:\n",
    "                X = get_region(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_region(protein, t_data, training_set_number)))\n",
    "        if arg == 'CLIP': \n",
    "            if X.size == 0:\n",
    "                X = get_cobinding(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_cobinding(protein, t_data, training_set_number)))\n",
    "        data_sources.append(arg)\n",
    "        \n",
    "    data_sources = ' '.join(data_sources)     \n",
    "    return data_sources, X\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create empty dictionary to create a table later.\n",
    "table_data = {}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ligter version where you can manually change set values\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "# Define a set of data from experiments to be used. Options include 'KMER', 'RNA', 'RG', 'CLIP'.\n",
    "experiment_set = ['KMER', 'RNA', 'RG', 'CLIP']\n",
    "score_list = []\n",
    "training_set_number = 0\n",
    "\n",
    "\n",
    "\n",
    "for protein in protein_list:\n",
    "\n",
    "    print (protein)\n",
    "\n",
    "    data_sources, X = load_data_sources(protein, 'train', training_set_number, *experiment_set)\n",
    "    y = get_class(protein,\"train\",training_set_number)\n",
    "    size = X.shape[2]\n",
    "    score = []\n",
    "\n",
    "    model = Sequential()\n",
    "    model.add(Conv1D(60,6, data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "    model.add(MaxPooling1D(pool_size=20, strides=1, padding='valid'))\n",
    "    model.add(Conv1D(60, 4, activation='relu'))\n",
    "    model.add(Dropout(0.2))\n",
    "    model.add(MaxPooling1D(pool_size=40, strides=1, padding='valid'))\n",
    "    model.add(Conv1D(60, 4, activation='relu'))\n",
    "    model.add(MaxPooling1D(pool_size=30, strides=1, padding='valid'))\n",
    "    model.add(Conv1D(60, 3, activation='relu'))\n",
    "    model.add(GlobalAveragePooling1D())\n",
    "    model.add(Dense(200, activation='relu'))\n",
    "    model.add(Dropout(0.1))\n",
    "    model.add(Dense(2, activation='sigmoid'))\n",
    "\n",
    "    model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "    checkpointer = ModelCheckpoint(filepath=\"models/\" + protein + \"_weights.hdf5\", verbose=0, save_best_only=True)\n",
    "    earlystopper = EarlyStopping(monitor='val_loss', patience=5, verbose=0)\n",
    "\n",
    "    X_train = X\n",
    "    y_train = y\n",
    "\n",
    "    metrics = model.fit(X_train, y_train, validation_split = 0.2, epochs=12, batch_size=200, verbose=0, callbacks=[earlystopper])\n",
    "\n",
    "    #Run predictions on test dataset.\n",
    "    data_sources, X_test = load_data_sources(protein, 'test', 0, *experiment_set)\n",
    "    y_test = get_class(protein,\"test\",training_set_number)\n",
    "\n",
    "\n",
    "    y_scores = model.predict(X_test)\n",
    "    y_scores = y_scores [:,0:1]\n",
    "    y_test = y_test [:,0:1]\n",
    "\n",
    "    score = roc_auc_score(y_test, y_scores)\n",
    "\n",
    "    print (score)\n",
    "    score_list.append(score)\n",
    "\n",
    "\n",
    "print(score_list)\n",
    "print (data_sources)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_max(s):\n",
    "    '''\n",
    "    highlight the maximum in a Series yellow.\n",
    "    '''\n",
    "    is_max = s == s.max()\n",
    "    return ['background-color: yellow' if v else '' for v in is_max]\n",
    "\n",
    "# Create a table to view results\n",
    "num_tags = [i for i in range (1,32)]\n",
    "table_data[data_sources] = np.round(score_list,2)\n",
    "table = pd.DataFrame(table_data, index = num_tags)\n",
    "table = table.T\n",
    "table.style.apply(highlight_max)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
